From: Ong Boon Leong <boon.leong.ong@intel.com>
Date: Wed, 31 Mar 2021 21:25:03 +0800
Subject: xdp: fix xdp_return_frame() kernel BUG throw for page_pool memory
 model
Patch-mainline: v5.12-rc7
Git-commit: 622d13694b5f048c01caa7ba548498d9880d4cb0
References: bsc#1109837

xdp_return_frame() may be called outside of NAPI context to return
xdpf back to page_pool. xdp_return_frame() calls __xdp_return() with
napi_direct = false. For page_pool memory model, __xdp_return() calls
xdp_return_frame_no_direct() unconditionally and below false negative
kernel BUG throw happened under preempt-rt build:

[  430.450355] BUG: using smp_processor_id() in preemptible [00000000] code: modprobe/3884
[  430.451678] caller is __xdp_return+0x1ff/0x2e0
[  430.452111] CPU: 0 PID: 3884 Comm: modprobe Tainted: G     U      E     5.12.0-rc2+ #45

Changes in v2:
 - This patch fixes the issue by making xdp_return_frame_no_direct() is
   only called if napi_direct = true, as recommended for better by
   Jesper Dangaard Brouer. Thanks!

Fixes: 2539650fadbf ("xdp: Helpers for disabling napi_direct of xdp_return_frame")
Signed-off-by: Ong Boon Leong <boon.leong.ong@intel.com>
Acked-by: Jesper Dangaard Brouer <brouer@redhat.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Thomas Bogendoerfer <tbogendoerfer@suse.de>
---
 net/core/xdp.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/net/core/xdp.c
+++ b/net/core/xdp.c
@@ -334,7 +334,8 @@ static void __xdp_return(void *data, str
 		xa = rhashtable_lookup(mem_id_ht, &mem->id, mem_id_rht_params);
 		page = virt_to_head_page(data);
 		if (xa) {
-			napi_direct &= !xdp_return_frame_no_direct();
+			if (napi_direct && xdp_return_frame_no_direct())
+				napi_direct = false;
 			page_pool_put_page(xa->page_pool, page, napi_direct);
 		} else {
 			put_page(page);
